### ISSUE TITLE
Turbo UI improvements

### ISSUE THREAD

**Issue by irilyth** [buttonmen-dev/buttonmen#2175]
Two small suggestions for improving the Turbo UI:

1. Could we put a red frame around the selected boxes? I think that would nicely match the visual hightlighting of selecting the dice.

2. If I'm keyboard-navigating, it'd be nice if hitting return in the text box of a selector would submit the form, just like it does after selecting a die.

**Comment by blackshadowshade** [comment]
I've discussed point 2 with Chaos before, and the current behaviour is consistent with what happens when we set swing values on the swing setting page.

If you happen to know how we can make the return key actually do the submit, I'd be happy to consider this, but it would need to be done in both places to keep the behaviour consistent.

**Comment by irilyth** [comment]
Hmm, if I type a number into the box on a Swing die selector screen, and hit return, it submits the form. Does that not happen for you?

**Comment by blackshadowshade** [comment]
Did you navigate there via keyboard?

**Comment by blackshadowshade** [comment]
Huh, on my local test site, it does submit, even when navigating by keyboard. I could have sworn that it didn't on Chaos's test instance.

**Comment by irilyth** [comment]
Sure; I browsed to http://54.160.177.210/ui/index.html, typed "61" to select game 61, hit return, hit tab a bunch until the cursor showed up in the swing die selection box, typed "4", hit return, it submitted the form.

**Comment by blackshadowshade** [comment]
Okay, the next question is *why* the return inside the turbo selector doesn't trigger the submit. I don't actually know the answer to that question. Do you?

**Comment by irilyth** [comment]
I don't think it should matter how you navigate into the box (whether by mouse or keyboard) if it does, that seems weird.

For things like selecting dice to attack with, it doesn't matter whether I select them by keyboard (tab and space) or by mouse -- if the last thing I did was select or de-select a die, hitting return submits the form.

**Comment by blackshadowshade** [comment]
I seem to remember that we needed to put a huge amount of logic into the keyboard handler to make it work the same as mouse selection.

**Comment by irilyth** [comment]
I don't know why hitting return inside the Turbo selector doesn't submit, but if I had to guess, I'd guess it's the same reason why hitting return after selecting a die didn't submit the form before -- it's all a bunch of Javascript, and not really a standard HTML form at that point, so the Javascript that interprets the keystrokes has to interpret return as submit in that context.

I don't know why it does in the context of clicking on a die, but not in the context of clicking in the Turbo box.

**Comment by blackshadowshade** [comment]
I'm pretty sure it has to do with the logic not yet being in the custom keyboard handler in Game.js. Digging through now to see what's going on there.

**Comment by irilyth** [comment]
Just to be extra clear: I don't think I've ever seen it matter how I selected the currently-selected thing, whether by clicking or by mousing. (Although I could be mis-remembering how it used to work before #180.) Once it's selected, hitting return consistently either does or doesn't submit the form (depending on whether it's a die or the Turbo selector).

In case this is useful info: Hitting return also works when the BPU button is selected (which you can only do by tabbing to it, since if you click on it, it submits :^), and when the Attack Type menu is selected (either by clicking or tabbing).

**Comment by blackshadowshade** [comment]
Okay, there's one hint here:

https://github.com/buttonmen-dev/buttonmen/blob/master/src/ui/js/Game.js, line 870, there's a comment that matches roughly what we want here:

`// install a keyboard handler so that hitting "return" while on
  // the attack type select submits the form (don't install any
  // mouse click or space bar behavior)
  Env.addClickKeyboardHandlers(
    attacktypeselect,
    null,
    null,
    Game.form
  );`

I'm guessing that I need to install this also for the turbo selector.

**Comment by blackshadowshade** [comment]
One question: what behaviour would you expect when pressing return on a turbo option selector? Would this reopen the option selector as usual, or would it submit the form?

**Comment by blackshadowshade** [comment]
Regarding the first point, I don't like the idea of the red box around the turbo selector much. I think that the parallelism signals that the die should be selectable/unselectable via clicking on the turbo selector.

**Comment by irilyth** [comment]
> One question: what behaviour would you expect when pressing return on a turbo option selector? Would this reopen the option selector as usual, or would it submit the form?

Hmm, what happens now on the Swing die selector page? Looks like it does NOT submit, so I'd say also not submit on the Turbo selector. Good question.

**Comment by irilyth** [comment]
> Regarding the first point, I don't like the idea of the red box around the turbo selector much. I think that the parallelism signals that the die should be selectable/unselectable via clicking on the turbo selector.

Yeah, I'm not 100% sure about it but I think it'd be nicer with the border. Would it be easy to demo it somewhere?

**Comment by irilyth** [comment]
Oh, I remember a third suggestion I had: Now that we've got the grayed-out selectors for non-selected dice, would it be better to keep the turbo selection area visible all the time, rather than having the whole area blink in and out as you select and unselect turbo dice? I think it might look better, and would be interested to see it if it's easy to demo.

**Comment by blackshadowshade** [comment]
So, I've been thinking more about the red border. What about changing the background colour of the selector to the player's mat color when it's active, instead of using a border? I think that would achieve the same aim of associating the die and the selector, without having the uncanny parallelism that I think has false implications.

**Comment by blackshadowshade** [comment]
Regarding the third suggestion, I think that the visual impact of this will be quite dependent on what we decide about the first point, so I'd like to defer discussion on that until we reach some consensus on the first point.

**Comment by irilyth** [comment]
> I think that would achieve the same aim of associating the die and the selector, without having the uncanny parallelism that I think has false implications.

I'm not sure what you mean here -- I think the specific thing the red border adds, beyond just the gray/not-gray we have now, is that implied parallelism. :^) "A selected thing has a red border", both for dice and for the Turbo selector. Can you say more about what you don't like about it?

> Regarding the third suggestion, I think that the visual impact of this will be quite dependent on what we decide about the first point, so I'd like to defer discussion on that until we reach some consensus on the first point.

Sure, that's fine.

**Comment by blackshadowshade** [comment]
Yeah, the implication is that you can click on the red-bordered thing to deselect it, but of course, that won't work.

I want to somehow visually keep the hierarchy of "the die is the only thing that can be selected/deselected", which is why I suggested making the turbo selector prominent in another way.

**Comment by irilyth** [comment]
Ah, I see, and I agree that makes sense too. Maybe gray and not-gray is good enough, then. If you want to try something to make it more prominent, go for it; otherwise, I think it's fine as-is.

If we don't change anything about the selected/unselected toggle, what do you think about that third point, having the Turbo area not appear and disappear?

**Comment by blackshadowshade** [comment]
Regarding the third point, I don't hold a strong view. I think it's nice that something happens  when you click a turbo die to prompt you that there's an extra action required, but the switch from grey to non-grey might be enough there too.

What do others think?

**Comment by irilyth** [comment]
Yeah, so my suggestion is:

* If having the Turbo area visible all the time, and just graying and un-graying the selector(s) is good enough, I like that best.
* If that's not a strong enough visual cue, I like the idea of having the background color change to the player's color, second best.
* If that's not strong enough either, or turns out to look ugly in practice, I think the current behavior is probably ok.

Anyone else have an opinion either way?

**Comment by AdmiralJota** [comment]
I strongly agree that the Turbo area should be visible at all times -- it feels weird to have it disappear completely when it's not being used.

That said, I don't really have a strong feeling about the best way to visually distinguish between whether a particular Turbo selector is active or not. I think that the current behavior of greying it out when it's not active would probably be sufficient (minus having the whole area disappear), but if folks wanted something more striking, like the red outline or changing the background color, I don't have any particular objections to that.

For the sake of deciding if they're "ugly in practice", here are some quick mockups...

Current-ish (without the disappearing):

![image](https://cloud.githubusercontent.com/assets/5848173/25568871/b99b93b2-2dda-11e7-87ed-3fe433e2bae5.png)

![image](https://cloud.githubusercontent.com/assets/5848173/25568864/910c1282-2dda-11e7-9c44-50cc0f1974ed.png)

Red outlines:

![image](https://cloud.githubusercontent.com/assets/5848173/25568942/fb7fc50a-2dda-11e7-80b2-b1e79e2b83a5.png)

Player's background color:

![image](https://cloud.githubusercontent.com/assets/5848173/25568969/5379edf8-2ddb-11e7-82e1-a0faa4a6568a.png)

Greyed background color:

![image](https://cloud.githubusercontent.com/assets/5848173/25568982/8bb4c094-2ddb-11e7-8af7-15722f7e26d0.png)

I think that my favorite of these is the last one, with the grey background, but I would be fine with any of them.

**Comment by cgolubi1** [comment]
Just based on your mockup, i like the grey background --- we use grey for inactive dice already, so that seems consistent.

**Comment by blackshadowshade** [comment]
Okay, looks good to me too. I'll look at getting a pull request in some time soon (unless someone else does) making the turbo area always visible, and using a grey background for the unselected turbo size specifier.

**Comment by irilyth** [comment]
I actually like the red border best - I really like the symmetry between the selected die and the selected input box. Hmm.

-Josh (from my phone)

**Comment by blackshadowshade** [comment]
Okay, I give up. I know how to make the selector always visible. However, I'm stumped by the CSS/JQuery needed to get the correct grey colouring of the selector.

**Comment by AdmiralJota** [comment]
Probably the best way would be to remove `style="background: none repeat scroll 0 0 #ffffff"` from the code that renders the HTML for the `turbo_subspan` element and add it to the CSS rule for `.turbo_subspan` instead. Then add a new CSS rule for `.turbo_subspan.disabled_turbo` that contains `background: none repeat scroll 0 0 #d7d7d7;`. (Or whichever shade of grey we pick -- I forget which one I used for the mockup.)

**Comment by blackshadowshade** [comment]
I think that this has been dealt with, so I'll close this. If anyone objects, they can reopen the issue.

**Comment by irilyth** [comment]
I don't think we ever got the submit-on-return thing working, did we?

**Comment by blackshadowshade** [comment]
Ah, good point.
